Pause when holding an arrow key

achu1998 opened the pull request:
Fix for #1609

1. A new property, _keysDown, was added to the viewer instance to track which keyboard keys are currently pressed. This enables the viewer to respond to continuous key presses, such as holding down an arrow key for smooth panning.

2. The onCanvasKeyDown and onCanvasKeyUp event handlers were implemented and registered. These handlers update the _keysDown property by setting the relevant key to true on keydown and to false on keyup, ensuring accurate tracking of user input.

3. The updateOnce function was modified to check the state of _keysDown for the arrow keys on each update cycle. If any arrow key is pressed, the viewer pans in the corresponding direction by a fixed amount (PAN_SPEED) per frame. This allows for responsive and continuous movement while keys are held.

4. The rest of the updateOnce function remains consistent with OpenSeadragon’s standard update logic, ensuring that the new keyboard panning integrates smoothly with existing features such as resizing, animation events, overlays, and navigator updates.

iangilman reviewed (changes requested):
Thank you for working on this! Looks like a good start, but there are a number of subtleties we need to address. I've noted a few below.

Also, I think it's important that you be able to have a consistent result when just tapping the key once. Because of that, I think what we need is for the initial keydown to move a certain distance, and then `updateOnce` to keep track of how long it's been since the last move, and then move periodically. In that way we move in larger chunks less often, rather than constantly with smaller chunks. Let me know if that makes sense!

achu1998 commented on src/viewer.js:
```diff
@@ -3992,6 +4000,23 @@ function doViewerResize(viewer, containerSize){
+    const PAN_SPEED = 0.01;
+
+    // Arrow key panning
+    if (viewer._keysDown['ArrowLeft']) {
```
The onCanvasKeydown logic has many other key actions, such as symbols and characters. Should I stick to just top, down, left, and right?

iangilman replied:
Yes, just the panning keys.

iangilman replied:
Actually, looking closer at `onCanvasKeyDown` there are others that should be addressed. In addition to the arrows, there are the WASD keys which also are for panning. There's also the + and - keys for zooming. Those should be supported in this new code as well.

iangilman commented on src/viewer.js:
```diff
@@ -3100,12 +3100,20 @@ function onCanvasContextMenu( event ) {
 
+
+/**
+ * Handles the keyup event on the viewer's canvas element.
```
Thank you for adding this comment. Please also add a `@private` line so the documentation builder knows this is private.

iangilman commented on src/viewer.js:
```diff
@@ -3990,8 +4006,57 @@ function doViewerResize(viewer, containerSize){
+
+function handleArrowKeys(viewer) {
+    // Use the viewer's configured pan amount
+    const pixels = viewer.pixelsPerArrowPress || 40;
```
Having the fallback here is nice defensive coding, but it would also mask the problem if for some reason that was missing; I think it would be better to leave it off so that we know about the problem if it happens.

iangilman reviewed (changes requested):
Coming along! 

> Yes, this makes sense. But just to clarify, if we move in larger chunks, doesn't that make a sudden big jump even though it is not that often?

Yes, but the smoothing in the panning animation should help with that.

That said, here's a different approach: pan every time in `updateOnce` but if the user lets the key up before a minimum distance has been travelled, continue behaving as if the key is held down for long enough to travel that distance. That might be a better approach.

Either way, the goal is to provide smooth movement, but also to have single taps produce a consistent movement distance.

A couple more notes: 

* It would be good to not have the key panning code duplicated between `onCanvasKeyDown` and `handleArrowKeys`. 
* Note that `onCanvasKeyDown` continues to be called periodically as long as the user is holding down the key. We don't want to be panning in response to that, because it'll create the same initial delay we're trying to fix here.

For those reasons, `onCanvasKeyDown` should just be setting the `_keysDown` and let `handleArrowKeys` do the actual panning. Some additional handling might be necessary to keep track of when the key was first pressed down (as discussed above in this comment). Note that non-panning/zooming keys should still be handled directly in `onCanvasKeyDown`.

You've done a great job of bringing the logic over to `handleArrowKeys`. One thing I see missing is the `event.preventDefault` (which is different from `canvasKeyDownEventArgs.preventDefaultAction`). That should continue to exist in `onCanvasKeyDown` (because that's when the event comes in) even when the rest of the panning key code has been cleaned out.

achu1998 commented:
> Coming along!
> 
> > Yes, this makes sense. But just to clarify, if we move in larger chunks, doesn't that make a sudden big jump even though it is not that often?
> 
> Yes, but the smoothing in the panning animation should help with that.
> 
> That said, here's a different approach: pan every time in `updateOnce` but if the user lets the key up before a minimum distance has been travelled, continue behaving as if the key is held down for long enough to travel that distance. That might be a better approach.
> 
> Either way, the goal is to provide smooth movement, but also to have single taps produce a consistent movement distance.
> 
> A couple more notes:
> 
> * It would be good to not have the key panning code duplicated between `onCanvasKeyDown` and `handleArrowKeys`.
> * Note that `onCanvasKeyDown` continues to be called periodically as long as the user is holding down the key. We don't want to be panning in response to that, because it'll create the same initial delay we're trying to fix here.
> 
> For those reasons, `onCanvasKeyDown` should just be setting the `_keysDown` and let `handleArrowKeys` do the actual panning. Some additional handling might be necessary to keep track of when the key was first pressed down (as discussed above in this comment). Note that non-panning/zooming keys should still be handled directly in `onCanvasKeyDown`.
> 
> You've done a great job of bringing the logic over to `handleArrowKeys`. One thing I see missing is the `event.preventDefault` (which is different from `canvasKeyDownEventArgs.preventDefaultAction`). That should continue to exist in `onCanvasKeyDown` (because that's when the event comes in) even when the rest of the panning key code has been cleaned out.

Yes, I will do this. I just don't want to mix this with the other review comments, as it is a bit confusing for me. I have fixed the others based on the review comments. If it is a thumbs-up for the changes, I will make the above changes on top of it, or can make a separate PR for this. Please let me know what you suggest. 

I have 2 clarification questions on the above

1. In the above comment, you mentioned to retain `event.preventDefault` in `onCanvasKeyDown`. It still exists there for each key case. Am I missing something here?

2. Note that non-panning/zooming keys should still be handled directly in `onCanvasKeyDown` - One of your other comments mentioned to move all the key functions like WASD to the new method `handleArrowKeys`. So if the new method `handleArrowKeys` has all the panning, WASD, +, - logic, `onCanvasKeyDown` will have only logic for 0, R, F, J, K?

iangilman commented on src/viewer.js:
```diff
@@ -4008,49 +4012,93 @@ function doViewerResize(viewer, containerSize){
+
+    if (!viewer.preventHorizontalPan) {
+        if (isDown('KeyA')) {
+            viewer.viewport.panBy(new OpenSeadragon.Point(-panDelta40.x, 0));
```
I see you have faithfully reproduced the 40 from the original code in `onCanvasKeyDown`. I believe that's actually a bug... the WASD keys and the arrow keys should behave the same. Let's use the regular `panDelta`.

iangilman replied:
Once you've done that, the arrow and WASD keys should be identical, so you can combine their code.

iangilman commented on src/viewer.js:
```diff
@@ -4008,49 +4012,93 @@ function doViewerResize(viewer, containerSize){
+    if (!viewer.preventVerticalPan) {
+        if (isDown('ArrowUp')) {
+            if (shift) {
+                viewer.viewport.zoomBy(1.1);
```
Working on this I'm seeing we have a few bugs from before that you can fix. Having the `zoomBy` inside the `!viewer.preventVerticalPan` is a bug... the pan prevention shouldn't prevent zooming.

iangilman commented on src/viewer.js:
```diff
@@ -3137,6 +3137,10 @@ function onCanvasKeyDown( event ) {
     this.raiseEvent('canvas-key', canvasKeyDownEventArgs);
 
+    if (!canvasKeyDownEventArgs.preventDefaultAction) {
+        this._keysDown[canvasKeyDownEventArgs.originalEvent.code] = true; // Mark this key as held down in the viewer's internal tracking object
```
This should probably be inside the if statement below; that way it is not engaged when those modifier keys are held down, maintaining the logic.

iangilman reviewed (changes requested):
Coming along! Totally reasonable to do this in parts. Thank you for sticking with it!

iangilman commented:
> Yes, I will do this. I just don't want to mix this with the other review comments, as it is a bit confusing for me. I have fixed the others based on the review comments. If it is a thumbs-up for the changes, I will make the above changes on top of it, or can make a separate PR for this. Please let me know what you suggest.

Let's keep going in this PR.

> In the above comment, you mentioned to retain event.preventDefault in onCanvasKeyDown. It still exists there for each key case. Am I missing something here?

I'm just saying that when you clean the panning and zooming code out of `onCanvasKeyDown` that we should keep the event.preventDefault.

> Note that non-panning/zooming keys should still be handled directly in onCanvasKeyDown - One of your other comments mentioned to move all the key functions like WASD to the new method handleArrowKeys. So if the new method handleArrowKeys has all the panning, WASD, +, - logic, onCanvasKeyDown will have only logic for 0, R, F, J, K?

Correct 😊

achu1998 commented:
@iangilman, I have made the changes according to the comments and implemented the enhancement as we discussed above. I have added a few helper functions as well to make the job simpler, as the code was becoming messier. Please review it and let me know if any other changes need to be made.

iangilman commented on src/viewer.js:
```diff
@@ -3990,8 +3956,91 @@ function doViewerResize(viewer, containerSize){
+    function handleZoom(codeList, factor) {
+        for (const code of codeList) {
+            if (isDown(code)) {
+                viewer.viewport.zoomBy(factor);
```
We need to disable the built in animation here as well. It should be `zoomBy(factor, null, true)`.

iangilman commented on src/viewer.js:
```diff
@@ -3990,8 +3956,91 @@ function doViewerResize(viewer, containerSize){
+        for (const code of codeList) {
+            if (isDown(code)) {
+                viewer.viewport.panBy(deltaPoint);
+                viewer.viewport.applyConstraints();
```
Also pass `true` here.

iangilman commented on src/viewer.js:
```diff
@@ -3990,8 +3956,91 @@ function doViewerResize(viewer, containerSize){
+        for (const code of codeList) {
+            if (isDown(code)) {
+                viewer.viewport.zoomBy(factor);
+                viewer.viewport.applyConstraints();
```
Pass `true` here.

iangilman commented on src/viewer.js:
```diff
@@ -3990,8 +3956,91 @@ function doViewerResize(viewer, containerSize){
+    }
+
+    // Use the viewer's configured pan amount
+    const pixels = viewer.pixelsPerArrowPress;
```
We need to adjust this to compensate for the fact we are now doing it every frame instead of every key press. Dividing by 10 seems good.

iangilman commented on src/viewer.js:
```diff
@@ -3990,8 +3956,91 @@ function doViewerResize(viewer, containerSize){
+    const panDelta = viewer.viewport.deltaPointsFromPixels(new OpenSeadragon.Point(pixels, pixels));
+
+    // Shift key state
+    const shift = isDown('Shift') || isDown('ShiftLeft') || isDown('ShiftRight');
```
This is an interesting situation. It's currently broken because those keys aren't part of the `isNavKey` list, so they are never marked as being down. That said, we don't want the same logic to work on them with the virtual key down. For that matter, I don't think we necessarily want to be able to hold down the W and switch between zooming and panning by pressing and letting up on the shift key.

This implies that maybe a cleaner approach would be to keep track of the actions rather than the keys. In other words, if the user presses W, we keep track of the fact that the "pan up" action is engaged. If they press shift W, we keep track of it as "zoom in". That makes the key down and key up logic a little more complicated but it simplifies this function.

achu1998 replied:
Currently corrected the logic to accommodate shift, so now it should not break. Will change the logic as you suggested.

iangilman commented on src/viewer.js:
```diff
@@ -3990,8 +3956,91 @@ function doViewerResize(viewer, containerSize){
+
+    // 1. Zoom with shift + up/down (WS or arrows)
+    if (shift) {
+        handleZoom(['ArrowUp', 'KeyW'], 1.1);
```
Again we need to compensate for the fact that we are doing it once per frame. 1.01 seems good to me.

iangilman commented on src/viewer.js:
```diff
@@ -3990,8 +3956,91 @@ function doViewerResize(viewer, containerSize){
+    // 1. Zoom with shift + up/down (WS or arrows)
+    if (shift) {
+        handleZoom(['ArrowUp', 'KeyW'], 1.1);
+        handleZoom(['ArrowDown', 'KeyS'], 0.9);
```
And 0.99 seems good here.

iangilman commented on src/viewer.js:
```diff
@@ -3990,8 +3956,91 @@ function doViewerResize(viewer, containerSize){
+    }
+
+    // 4. Zoom in/out (+ =, - _)
+    handleZoom(['Equal', 'NumpadAdd'], 1.1);
```
Same note here for the zoom values.

iangilman commented on src/viewer.js:
```diff
@@ -3990,8 +3956,91 @@ function doViewerResize(viewer, containerSize){
+    // If a key is held down or virtually held, increment its pan distance by panStep.
+    // Once the minimum pan distance is reached, stop virtually holding the key and reset the counter.
+
+    for (let code in viewer._keysDown) {
```
This can go inside of `handleArrowKeys`.

iangilman commented on src/viewer.js:
```diff
@@ -3990,8 +3956,91 @@ function doViewerResize(viewer, containerSize){
+    for (let code in viewer._keysDown) {
+        if (viewer._keysDown[code] || viewer._keyVirtuallyHeld[code]) {
+            // pan by step
+            viewer._keyPanDistance[code] += viewer.panStep;
```
This would just become a `++` since we are counting frames.

iangilman reviewed (changes requested):
Thank you for your patience. There's a lot to think about in here! Getting the UX right reveals surprising complexity! I hope you don't mind continuing this journey.

achu1998 commented on src/viewer.js:
```diff
@@ -3990,8 +3956,91 @@ function doViewerResize(viewer, containerSize){
+            if (isDown(code)) {
+                viewer.viewport.panBy(deltaPoint);
+                viewer.viewport.applyConstraints();
+                viewer._keyPanDistance[code] += viewer.pixelsPerArrowPress;
```
So just moving it outside handleNavKeys will do it?

iangilman replied:
You should just get rid of this logic inside `handlePan`. It's a duplicate of the logic you now have at the top of `handleNavKeys`.

iangilman commented on src/viewer.js:
```diff
@@ -3957,23 +3960,38 @@ function doViewerResize(viewer, containerSize){
+function handleNavKeys(viewer) {
+
+    // Iterate over all keys to handle keyboard-based panning.
+    // If a key is held down or virtually held, increment its pan distance by panStep.
```
This comment is no longer in line with what the code actually does.

iangilman commented on src/viewer.js:
```diff
@@ -3957,23 +3960,38 @@ function doViewerResize(viewer, containerSize){
+
+    for (let code in viewer._navKeysDown) {
+        if (viewer._navKeysDown[code] || viewer._navKeysVirtuallyHeld[code]) {
+            // pan by step
```
Likewise this comment. We don't really need this one anyway.

iangilman commented on src/viewer.js:
```diff
@@ -3957,23 +3960,38 @@ function doViewerResize(viewer, containerSize){
+            viewer._navKeyFrames[code]++;
+            if (viewer._navKeyFrames[code] >= viewer._minNavKeyFrames) {
+                viewer._navKeysVirtuallyHeld[code] = false;
+                viewer._navKeyFrames[code] = 0;
```
We shouldn't be resetting the _navKeyFrames here. On the one hand it's unnecessary (since it gets reset when the key gets pressed) but also it's problematic (because it resets the count if the key is still held down).

iangilman commented on src/viewer.js:
```diff
@@ -3122,94 +3163,22 @@ function onCanvasKeyDown( event ) {
+
+        if (isNavKey(code)) {
+            this._navKeysDown[code] = true; // Mark this key as held down in the viewer's internal tracking object
+            this._navKeyFrames[code] = 0; // Reset key frames
```
This is good; I think this is the right place for resetting the _navKeyFrames. However, onCanvasKeyDown gets called on key repeat (if the user is holding the key down), and we don't want to reset in that case. Check to see if `_navKeysDown[code]` is already true and don't reset the frames in that case.

iangilman reviewed (changes requested):
Coming along nicely! How do you feel about how it's behaving?

I had another thought: we're not going to get key events if the viewer loses focus, so at that point we should assume the user has let up on all of the keys. Please add to `onCanvasBlur` some code to set all the _navKeysDown and _navKeysVirtuallyHeld to false.

iangilman commented:
Looking great so far! 😊

iangilman commented on src/viewer.js:
```diff
@@ -3114,25 +3126,74 @@ function isNavKey(code) {
+            this._navActionVirtuallyHeld[action] = true;
+        }
+    } else {
+        this._activeActions.zoomIn = false;
```
What's the purpose of this section?

iangilman commented on src/viewer.js:
```diff
@@ -246,6 +246,23 @@ $.Viewer = function( options ) {
+    this._minNavActionFrames = 10;      // minimum pan distance per tap or key press
+
+    this._activeActions = { // variable to keep track of currently pressed action
+    // Basic arrow key panning (no modifiers)
```
Everything inside the object here should be indented once.

iangilman reviewed (changes requested):
This is looking much cleaner!

Looks like we have one issue I hadn't anticipated... what if the user holds down shift and then W, but then lets up the shift key before letting up the W? Then `onCanvasKeyUp` will turn off `panUp` instead of `zoomIn`, and the viewer will continue zooming. I assume that's what the `else` clause in onCanvasKeyUp is for. I don't think it's quite right though... it's too blanket. Also it doesn't cover what happens if the user starts holding down W and then holds down shift before letting up the W (no matter how rare that occurrence may be).

I'm not sure what the right approach is, but maybe the thing to do is have 'getActiveActionKey` take the code and shift params directly (instead of being passed as the event) and then you can handle both the shift and non-shift versions in onCanvasKeyUp.

Other than that, this is looking beautiful!

achu1998 commented on src/viewer.js:
```diff
@@ -3097,7 +3115,91 @@ function onCanvasContextMenu( event ) {
+ * Returns true if the key triggers panning (arrow keys, WASD) or zooming (+, -),
+ * otherwise returns false.
+ */
+function isNavKey(code) {
```
Yep, makes sense. Sorry about that.

iangilman commented:
> I have a question here. If we press shift + W, it will zoom in, and then if we release shift, it should continue panning upwards like the current implementation, right? For this, we should know what all keys are still pressed, right? To achieve this, we should bring back the _navKeysDown variable to keep track of which keys are pressed. Isn't this again going back to the previous version of code (logic without _activeActions)? Please correct me if i am wrong here.

I don't think we need to keep track of whether the shift key is held down. For this scenario, what's important is the state of the W key. The only time we need to care about the shift key is during the initial press.

Here's my proposed change:

```
function onCanvasKeyUp(event) {
    function processCombo(code, shift) {
        const action = getActiveActionFromKey(code, shift);
        if (action && this._activeActions[action]) {
            this._activeActions[action] = false;
            if (this._navActionFrames[action] < this._minNavActionFrames) {
                this._navActionVirtuallyHeld[action] = true;
            }
        }
    }

    // We don't know if the shift key was held down originally, so we check them both.
    const code = event.originalEvent.code;
    processCombo(code, true);
    processCombo(code, false);
}
```

In this way, we properly mark anything the W key could do as being let up. 

Note that this depends on `getActiveActionKey` taking in two arguments.

Does that make sense?

iangilman reviewed (approved):
Beautiful! This is looking really nice. Thank you for the excellent comments in the code as well!

And thank you for sticking through this process... I know it may have been more than you thought you were taking on, but you did a great job of it!

iangilman commented:
@achu1998 Wonderful, thank you for the offer! You've been making great contributions to this project!
